### PR TITLE
Add support for Mermaid diagrams

### PR DESCRIPTION
**Pull request by radupotop**
Added support for Mermaid diagrams: https://mermaid.js.org/intro/ 

* supports rendering to PNG only for now 
  * due to issues with rendering SVGs for some types of diagrams 
  * e.g. Sequence diagrams render fine but Flow charts and Class diagrams fail to render labels
* Is a bit slower than I'd like when rendering a page full of diagrams (takes about 5 seconds to render 7 diagrams), but maybe this is something which can be improved in the future.
  * `mmdc` supports rendering multiple diagrams at once

### CODE REVIEW

**Comment by introt** [comment]
Hi, thanks for the PR! Two things to note before further review:

1. We use tabs, not spaces. You can ask, but you can not complain, for it is the will of our [Almighty Tallest](https://zim.fandom.com/wiki/Almighty_Tallest).
2. Plugins and all of their options must be documented in the manual or you'll make our test suite upset.

```

2024-05-06T20:52:13.7724421Z ======================================================================
2024-05-06T20:52:13.7725377Z FAIL: testIndenting (tests.package.TestCoding)
2024-05-06T20:52:13.7726001Z ----------------------------------------------------------------------
2024-05-06T20:52:13.7726497Z Traceback (most recent call last):
2024-05-06T20:52:13.7733980Z   File "/home/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/package.py", line 167, in testIndenting
2024-05-06T20:52:13.7735096Z     self.assertFalse(' ' in indent, 'Indenting should use tabs - file: %s line %s' % (file, lineno))
2024-05-06T20:52:13.7736155Z AssertionError: True is not false : Indenting should use tabs - file: zim/plugins/mermaiddiagrameditor.py line 18
2024-05-06T20:52:13.7736736Z 
2024-05-06T20:52:13.7736873Z ======================================================================
2024-05-06T20:52:13.7737304Z FAIL: runTest (tests.plugins.TestPluginClasses)
2024-05-06T20:52:13.7738215Z ----------------------------------------------------------------------
2024-05-06T20:52:13.7738853Z Traceback (most recent call last):
2024-05-06T20:52:13.7739613Z   File "/home/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/plugins.py", line 58, in runTest
2024-05-06T20:52:13.7740534Z     self.assertIn(rellink, pluginindex, 'Missing links "%s" in manual/Plugins.txt' % rellink)
2024-05-06T20:52:13.7751229Z AssertionError: '+Mermaid Diagram Editor' not found in 'Content-Type: text/x-zim-wiki\nWiki-Format: zim 0.6\n\n====== Plugins ======\n\nThe following plugins are standard included with zim:\n\n* [[+Arithmetic|Arithmetic]]\n* [[+Attachment Browser|Attachment Browser]]\n* [[+BackLinks Pane|BackLinks Pane]]\n* [[+BookmarksBar|BookmarksBar]]\n* [[+Command Palette|Command Palette]]\n* [[+Diagram Editor|Diagram Editor]]\n* [[+Distraction Free Editing|Distraction Free Editing]]\n* [[+Ditaa Editor|Ditaa Editor]]\n* [[+Equation Editor|Equation Editor]]\n* [[+GNU R Plot Editor|GNU R Plot Editor]]\n* [[+Gnuplot Editor|Gnuplot Editor]]\n* [[+Indexed Full Text Search|Indexed Full Text Search]]\n* [[+Insert Screenshot|Insert Screenshot]]\n* [[+Insert Symbol|Insert Symbol]]\n* [[+Inline Calculator|Inline Calculator]]\n* [[+Journal|Journal]]\n* [[+Link Map|Link Map]]\n* [[+Line Sorter|Line Sorter]]\n* [[+Log events with Zeitgeist|Log events with Zeitgeist]]\n* [[+macOS Menubar|macOS Menubar]]\n* [[+PageIndex|PageIndex]]\n* [[+PathBar|PathBar]]\n* [[+Print to Browser|Print to Browser]]\n* [[+Quick Note|Quick Note]]\n* [[+Score Editor|Score Editor]]\n* [[+Sequence Diagram Editor|Sequence Diagram Editor]]\n* [[+Source View|Source View]]\n* [[+Spell Checker|Spell Checker]]\n* [[+Table Editor|Table Editor]]\n* [[+Table Of Contents|Table Of Contents]]\n* [[+Tags|Tags]]\n* [[+Task List|Task List]]\n* [[+ToolBar|ToolBar]]\n* [[+Tray Icon|Tray Icon]]\n* [[+Version Control|Version Control]]\n* [[+WindowTitleEditor|Window Title Editor]]\n\nPlugins can be enabled and configured in the [[Help:Preferences|preferences dialog]]. Several plugins have additional dependencies, like specific external programs, that are not required for the core functionality of zim. If one or more dependencies of a plugin are not fulfilled, these are marked red.\n\n===== Custom Plugins =====\nMore plugins can be found here: https://github.com/jaap-karssenberg/zim-wiki/wiki/Plugins\n\n===== Writing Plugins =====\nIf you are looking for information on **writing plugins** please download the source package and have a look at the module documentation of the "\'\'zim.plugins\'\'" python module as well as the "PLUGIN_WRITING.md" file.\n' : Missing links "+Mermaid Diagram Editor" in manual/Plugins.txt
2024-05-06T20:52:13.7760212Z 
2024-05-06T20:52:13.7760440Z ----------------------------------------------------------------------
2024-05-06T20:52:13.7760843Z Ran 1013 tests in 44.408s
2024-05-06T20:52:13.7761026Z 
2024-05-06T20:52:13.7761198Z FAILED (failures=2, skipped=34, expected failures=21)
```

**Comment by radupotop** [comment]
Related mermaid-cli issue here: https://github.com/mermaid-js/mermaid-cli/issues/691

**Comment by radupotop** [comment]
> Hi, thanks for the PR! Two things to note before further review:
> 
>     1. We use tabs, not spaces. You can ask, but you can not complain, for it is the will of our [Almighty Tallest](https://zim.fandom.com/wiki/Almighty_Tallest).
> 
>     2. Plugins and all of their options must be documented in the manual or you'll make our test suite upset.
> 

Thanks for your review, can you please take another look?

File: zim/plugins/mermaiddiagrameditor.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+		self.mmdfile.write(text)
+		self.imgfile = LocalFile(self.mmdfile.path[:-4] + self.imagefile_extension)
+		logger.debug('Writing diagram to temp file: %s', self.imgfile)
+
+		# Call mmdc
+		try:
+			diag = Application(self.diagcmd)
+			diag.run((self.mmdfile, '-o', self.imgfile))
+		except ApplicationError:
+			return None, None  # Sorry, no log
```
Comment (introt) [inline zim/plugins/mermaiddiagrameditor.py]: I know this is copied from the other generators, but I'd much prefer `except ApplicationError as e:` & `logger.debug('Generating diagram failed with error: %s`, e)

**Review by introt — APPROVED** [review]
Thanks, looks good to me!

File: zim/plugins/mermaiddiagrameditor.py (hunk 1 above)
Comment (radupotop) [reply zim/plugins/mermaiddiagrameditor.py]: Done now

**Review by introt — APPROVED** [review]

**Comment by tridian-tn** [comment]
Hi @radupotop. I was browsing the recent Zim PRs and saw this one, which is all very exciting: it'll be brilliant to have mermaid support baked into Zim. Did you get it working on Windows? No worries if not. I only ask because I didn't notice this PR in September and I made fixes to another plugin (https://github.com/zim-plugins/insert_mermaid/pull/3/commits) to get it working. A pretty basic tweak to use "mmdc.cmd" if (and only if) running on Windows. I'm happy to make the change if needed. Cheers!

**Comment by radupotop** [comment]
Hey @tridian-tn I did not test on Windows since I don't have it on any of my machines, if you can provide a patch that would be awesome. Thanks.

**Pull request merged**